src/shogun/base/progress.h

```diff
@@ -0,0 +1,434 @@
+/*
+* Written (W) 2017 Giovanni De Toni
+*/
+
+#ifndef __SG_PROGRESS_H__
+#define __SG_PROGRESS_H__
+
+#include <iterator>
+#include <memory>
+#include <string>
+
+#include <shogun/base/init.h>
+#include <shogun/base/range.h>
+#include <shogun/io/SGIO.h>
+#include <shogun/lib/Time.h>
+#include <shogun/mathematics/Math.h>
+
+#if WIN32
+#include <windows.h>
+#else
+#include <sys/ioctl.h>
+#include <unistd.h>
+#endif
+
+namespace shogun
+{
+
+	/** Possible print modes */
+	enum SG_PRG_MODE
+	{
+		ASCII,
+		UTF8
+	};
+
+	/** @class Printer that display the progress bar
+	*
+	*/
+	class ProgressPrinter
+	{
+	public:
+		/** Creates a ProgressPrinter instance given an SGIO object
+		*
+		* @param    io  SGIO object
+		*/
+		ProgressPrinter(
+		    const SGIO& io, float64_t max_value, float64_t min_value)
+		    : m_io(io), m_max_value(max_value), m_min_value(min_value),
+		      m_prefix("PROGRESS: "), m_mode(UTF8), m_last_progress(0),
+		      m_last_progress_time(0), m_progress_start_time(0)
+		{
+		}
+		ProgressPrinter(
+		    const SGIO& io, float64_t max_value, float64_t min_value,
+		    const std::string& prefix)
+		    : m_io(io), m_max_value(max_value), m_min_value(min_value),
+		      m_prefix(prefix), m_mode(UTF8), m_last_progress(0),
+		      m_last_progress_time(0), m_progress_start_time(0)
+		{
+		}
+		ProgressPrinter(
+		    const SGIO& io, float64_t max_value, float64_t min_value,
+		    const SG_PRG_MODE mode)
+		    : m_io(io), m_max_value(max_value), m_min_value(min_value),
+		      m_prefix("PROGRESS: "), m_mode(mode), m_last_progress(0),
+		      m_last_progress_time(0), m_progress_start_time(0)
+		{
+		}
+		ProgressPrinter(
+		    const SGIO& io, float64_t max_value, float64_t min_value,
+		    const std::string& prefix, const SG_PRG_MODE mode)
+		    : m_io(io), m_max_value(max_value), m_min_value(min_value),
+		      m_prefix(prefix), m_mode(mode), m_last_progress(0),
+		      m_last_progress_time(0), m_progress_start_time(0)
+		{
+		}
+		~ProgressPrinter()
+		{
+		}
+
+		/** Print the progress bar */
+		void print_progress(float64_t current_value) const
+		{
+			// Check if the progress was enabled
+			if (!m_io.get_show_progress())
+				return;
+
+			if (m_max_value <= m_min_value)
+				return;
+
+			// Check for terminal dimension. This is for provide
+			// a minimal resize functionality.
+			set_screen_size();
```

Comment:
This method is run each time we print the progress bar to account for the terminal size, which can change (for instance, if the user resizes or expands the window). However, I'm not so sure this is worth it (since it doesn't work so well). Should I remove this line? @vigsterkr